PlayFabSDK/source/Qos/PlayFabQosApi.cs

```diff
@@ -0,0 +1,163 @@
+        private readonly Action<string> _logAction;
+
+        public PlayFabQosApi(Action<string> logAction = null)
+        {
+            _logAction = logAction;
+        }
+
+        public async Task<QosResult> GetQosResultAsync(int timeoutMs = DefaultTimeoutMs)
+        {
+            QosResult result = await GetResultAsync(timeoutMs);
+            if (result.ErrorCode != 0)
+            {
+                return result;
+            }
+
+            bool eventSent = await SendResultsToPlayFab(result);
+            if (!eventSent)
+            {
+                result.ErrorCode = -3;
+            }
+
+            return result;
+        }
+
+        private async Task<QosResult> GetResultAsync(int timeoutMs)
+        {
+            var result = new QosResult();
+
+            if (!PlayFabClientAPI.IsClientLoggedIn())
+            {
+                LogQos("Client is not logged in");
+                result.ErrorCode = -1;
+                return result;
+            }
+
+            // get datacenter map (call thunderhead)
+            await GetQoSServerList();
+
+            int serverCount = _dataCenterMap.Count;
+            if (serverCount < 0)
+            {
+                result.ErrorCode = -2;
+                return result;
+            }
+
+            // ping servers
+            result.RegionResults = await GetSortedRegionLatencies(timeoutMs);
+            result.ErrorCode = 0;
+            return result;
+        }
+
+        private async Task GetQoSServerList()
+        {
+            if (_dataCenterMap.Count > 0)
+            {
+                // If the dataCenterMap is already initialized, return
+                return;
+            }
+
+            var request = new ListQosServersRequest();
+            PlayFabResult<ListQosServersResponse> response = await PlayFabMultiplayerAPI.ListQosServersAsync(request);
+            
+            if (response?.Error != null)
+            {
+                LogQos("Could not get the server list from thunderhead\n Error : " + response.Error.ErrorMessage);
+            }
+
+            foreach (QosServer qosServer in response.Result.QosServers)
+            {
+                if (!qosServer.Region.HasValue)
+                {
+                    continue;
+                }
+
+                _dataCenterMap[qosServer.Region.Value] = qosServer.ServerUrl;
+            }
+        }
+
+        private async Task<List<QosRegionResult>> GetSortedRegionLatencies(int timeoutMs)
+        {
+            List<Task<QosRegionResult>> asyncPingResults = _dataCenterMap.Select(async kvp =>
+            {
+                var regionPinger = new RegionPinger(kvp.Value, kvp.Key);
+                return await regionPinger.PingAsync(timeoutMs);
+            }).ToList();
+
+            await Task.WhenAll(asyncPingResults);
+            return asyncPingResults.Select(t => t.Result).OrderBy(t => t.LatencyMs).ToList();
+        }
+
+        private async Task<bool> SendResultsToPlayFab(QosResult result)
```

Comment:
Design wise, is the bool return on this meaningful?
It seems like it could just be a fire-and-forget that executes a task async, without blocking the caller.

Reply:
I think it makes sense to not block the client on emitting metrics. Updated